galvasr2/zero_shot_classification_pipeline.py

```diff
@@ -0,0 +1,122 @@
+import pandas as pd
+from flair.data import Sentence
+from flair.models import SequenceTagger
+import logging
+from absl import flags
+from multiprocessing.pool import ThreadPool
+from pyspark.sql import SparkSession
+from tqdm import tqdm
+
+from galvasr2.align.spark.align_lib import (
+    load_audio_id_text_id_mapping,
+    load_transcripts,
+)
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_string(
+    "input_transcription_index",
+    "gs://the-peoples-speech-west-europe/archive_org/Mar_7_2021/CC_BY_SA_EXPANDED_LICENSES_FILTERED_ACCESS.jsonl.gz",
+    "Ubication of the path with the transcription index data",
+)
+flags.DEFINE_string(
+    "input_transcription_data",
+    "gs://the-peoples-speech-west-europe/archive_org/Mar_7_2021/CC_BY_SA_EXPANDED_LICENSES_FILTERED_ACCESS",
+    "Ubication of the path with the transcription data",
+)
+flags.DEFINE_string(
+    "num_rows",
+    10,
+    "Number of transcripts to analyze",
+)
+
+
+class ZSClassification:
+    def __init__(self, num_rows: int):
+        self.num_rows = num_rows
+
+    def read_transcription_data(
+        self,
+        spark: SparkSession,
+        data_trans_index: str,
+        data_trans: str,
+    ) -> pd.DataFrame:
+        # spark.sparkContext.setLogLevel("INFO") # "ALL" for very verbose logging
+        logging.getLogger("py4j").setLevel(logging.ERROR)
+        catalogue_df = load_audio_id_text_id_mapping(spark, data_trans_index)
+        training_sample_rows = catalogue_df.collect()
+
+        # Comment this out to load everything. It might takes ~15 minute, in my experience, on an 8 core machine.
+        training_sample_rows = training_sample_rows[: self.num_rows]
+        transcripts_df = load_transcripts(spark, data_trans, training_sample_rows)
+        transcripts_pdf = transcripts_df.toPandas()
+        return transcripts_pdf
+
+    def get_text_classification(transcripts_pdf: pd.DataFrame) -> pd.DataFrame:
+        zsl = text.ZeroShotClassifier()
+        labels = [
+            "politics",
+            "elections",
+            "sports",
+            "films",
+            "television",
+            "artificial intelligence",
+            "food",
+            "healthy",
+            "Information technology",
+            "financials",
+            "communication services",
+        ]
+
+        def get_top_class(row):
+            try:
+                row = row[:20000]
```

Comment:
It looks like you are truncating the string you are classifying. How come? Do you get out of memory errors if the string is too large? If so, can you comment that?

Reply:
This is a little tricky, there some transcriptions that are very large, the problem is not about memory errors or something similar, is about time, I think that with the batch size that you comment I could fix that

Reply:
Another reason, was because in theory, if you read, I don't know, 30% of a document, you can infer with high accuracy the most accurate category

Reply:
"Prediction Time and Batch Size
The predict method can accept a large list of documents. Documents are automatically split into batches based on the batch_size parameter, which can be increased to speed up predictions.

Note also that the predict method of ZeroShotClassifier generates a separate NLI prediction for each label included in the labels parameter. As len(labels) and the number of documents fed to predict increases, the prediction time will also increase. You can speed up predictions by increasing the batch_size. The default batch_size is currently set conservatively at 8:

- Predicting 800 topics for a single document takes ~26 seconds on a TITAN V GPU using batch_size=1
- Predicting 800 topics for a single document takes less than 2 seconds on a TITAN V GPU using batch_size=64¶
""
From the official documentation 

You were completely right, sorry

Reply:
It looks like increasing the batch_size is not a magic bullet. I guess you cannot increase it beyond the number of zero shot classification categories we use, if I am understanding correctly. I'm guessing this is a transformer model. Still, it may be worthwhile to set batch_size=len(labels). It you can evaluate the speedup on a small amount of data, that would be great. If it is too hard, I don't mind if you just leave this alone. 38 hours on a rarely run job is not horrible.